<a href="https://colab.research.google.com/github/LeeSeungwon89/Kaggle_Dacon_Practice/blob/main/4.%20Categorical_Feature_Encoding_Challenge_improve_model_performance2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install kaggle
from google.colab import files
files.upload()

In [3]:
ls -1ha kaggle.json

kaggle.json


In [4]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

# Permission Warning이 발생하지 않도록 해줍니다.
!chmod 600 ~/.kaggle/kaggle.json

# 참가한 대회 리스트를 확인합니다.
# !kaggle competitions list

In [5]:
!kaggle competitions download -c cat-in-the-dat

 42% 9.00M/21.2M [00:00<00:00, 69.4MB/s]
100% 21.2M/21.2M [00:00<00:00, 123MB/s] 


In [6]:
!ls

cat-in-the-dat.zip  kaggle.json  sample_data


In [7]:
!unzip cat-in-the-dat.zip

Archive:  cat-in-the-dat.zip
  inflating: sample_submission.csv   
  inflating: test.csv                
  inflating: train.csv               


# **1. 모델 성능 개선 절차**

이번 챕터에서는 XGBoost, LightGBM 모델의 하이퍼파라미터를 하나씩 추가하면서 최적 하이퍼파라미터를 도출해 보겠습니다. 아울러 학습 세트와 검증 세트의 비율을 0.1로 낮추고 다시 최적 하이퍼파라미터를 도출하여 모델 성능 차이를 확인해 보겠습니다.

# **2. 피처 엔지니어링**

## **2.1. 데이터 준비**

데이터를 준비하겠습니다. 'id' 피처의 고윳값 개수가 레코드 개수와 일치하므로 'id' 피처를 인덱스로 삼겠습니다.

In [8]:
import numpy as np
import pandas as pd

# 최대 행렬 수를 설정합니다.
pd.set_option('display.max_column', 50)
pd.set_option('display.max_rows', 50)

# 데이터를 읽습니다. 'id' 피처는 인덱스로 지정합니다.
train = pd.read_csv('train.csv', index_col='id')
test = pd.read_csv('test.csv', index_col='id')
submission = pd.read_csv('sample_submission.csv', index_col='id')

훈련 세트와 테스트 세트를 위아래로 쌓고 데이터 정보를 확인하겠습니다.

In [9]:
all_data = pd.concat([train, test], ignore_index=False)
all_data

bin_0  bin_1  bin_2 bin_3 bin_4  nom_0      nom_1    nom_2  \
id                                                                   
0           0      0      0     T     Y  Green   Triangle    Snake   
1           0      1      0     T     Y  Green  Trapezoid  Hamster   
2           0      0      0     F     Y   Blue  Trapezoid     Lion   
3           0      1      0     F     Y    Red  Trapezoid    Snake   
4           0      0      0     F     N    Red  Trapezoid     Lion   
...       ...    ...    ...   ...   ...    ...        ...      ...   
499995      0      0      0     F     N  Green     Square     Lion   
499996      1      0      0     F     Y  Green  Trapezoid     Lion   
499997      0      1      1     T     Y  Green  Trapezoid     Lion   
499998      1      0      0     T     Y   Blue       Star  Hamster   
499999      0      0      0     T     Y  Green       Star     Lion   

             nom_3     nom_4      nom_5      nom_6      nom_7      nom_8  \
id                                                                         
0          Finland   Bassoon  50f116bcf  3ac1b8814  68f6ad3e9  c389000ab   
1           Russia     Piano  b3b4d25d0  fbcb50fc1  3b6dd5612  4cd920251   
2           Russia  Theremin  3263bdce5  0922e3cb8  a6a36f527  de9c9f684   
3           Canada      Oboe  f12246592  50d7ad46a  ec69236eb  4ade6ab69   
4           Canada      Oboe  5b0f5acd5  1fe17a1fd  04ddac2be  cb43ab175   
...            ...       ...        ...        ...        ...        ...   
499995      Canada  Theremin  6c90f81cd  a406a5f12  d02a6b0ba  9e4b23160   
499996       China     Piano  2d61990e2  520806ce2  d2d8eabdb  cfbd87ed0   
499997      Canada      Oboe  488406659  28645754b  22831fffe  1108bcd6c   
499998  Costa Rica   Bassoon  f9d17bb93  2eadb68c5  fef807a3e  606ac930b   
499999       India     Piano  a5c276589  c135cdf77  a8f53c631  4ea576eb6   

            nom_9  ord_0        ord_1        ord_2 ord_3 ord_4 ord_5  day  \
id                                                                          
0       2f4cb3d51      2  Grandmaster         Cold     h     D    kr    2   
1       f83c56c21      1  Grandmaster          Hot     a     A    bF    7   
2       ae6800dd0      1       Expert     Lava Hot     h     R    Jc    7   
3       8270f0d71      1  Grandmaster  Boiling Hot     i     D    kW    2   
4       b164b72a7      1  Grandmaster     Freezing     a     R    qP    7   
...           ...    ...          ...          ...   ...   ...   ...  ...   
499995  acc31291f      1       Novice     Lava Hot     j     A    Gb    1   
499996  eae3446d0      1  Contributor     Lava Hot     f     S    Ed    2   
499997  33dd3cf4b      1       Novice  Boiling Hot     g     V    TR    3   
499998  d4cf587dd      2  Grandmaster  Boiling Hot     g     X    Ye    2   
499999  2d610f52c      2       Novice     Freezing     l     J    ex    2   

        month  target  
id                     
0           2     0.0  
1           8     0.0  
2           2     0.0  
3           1     1.0  
4           8     0.0  
...       ...     ...  
499995      3     NaN  
499996      2     NaN  
499997      1     NaN  
499998      1     NaN  
499999      2     NaN  

[500000 rows x 24 columns]

In [10]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500000 entries, 0 to 499999
Data columns (total 24 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   bin_0   500000 non-null  int64  
 1   bin_1   500000 non-null  int64  
 2   bin_2   500000 non-null  int64  
 3   bin_3   500000 non-null  object 
 4   bin_4   500000 non-null  object 
 5   nom_0   500000 non-null  object 
 6   nom_1   500000 non-null  object 
 7   nom_2   500000 non-null  object 
 8   nom_3   500000 non-null  object 
 9   nom_4   500000 non-null  object 
 10  nom_5   500000 non-null  object 
 11  nom_6   500000 non-null  object 
 12  nom_7   500000 non-null  object 
 13  nom_8   500000 non-null  object 
 14  nom_9   500000 non-null  object 
 15  ord_0   500000 non-null  int64  
 16  ord_1   500000 non-null  object 
 17  ord_2   500000 non-null  object 
 18  ord_3   500000 non-null  object 
 19  ord_4   500000 non-null  object 
 20  ord_5   500000 non-null  object 
 21  day     50

In [11]:
def feature_table(df):
    # 데이터 크기를 출력합니다.    
    print(df.shape)
    # 피처 이름과 유형을 담은 데이터프레임을 생성합니다.   
    feature_description = pd.DataFrame(df.dtypes, columns=['type'])
    # 피처 이름 피처가 인덱스이므로 인덱스를 리셋하여 새 인덱스를 만듭니다.
    feature_description = feature_description.reset_index()
    # 피처 이름을 값으로 가진 피처명은 'index'입니다. 'feature'로 변환합니다.
    feature_description = feature_description.rename(columns={'index': 'feature'})
    # 각 피처가 가진 첫 번째 ~ 다섯 번째 값을 출력합니다.
    feature_description['first_value'] = df.loc[0].values
    feature_description['second_value'] = df.loc[1].values
    feature_description['third_value'] = df.loc[2].values
    feature_description['fourth_value'] = df.loc[3].values
    feature_description['fifth_value'] = df.loc[4].values
    # 고윳값 개수를 산출합니다.
    feature_description['eigenvalue_count'] = df.nunique().values
    # 결측치 개수를 산출합니다.
    feature_description['null_count'] = df.isnull().sum().values

    return feature_description

feature_table(train)

(300000, 24)


feature    type  first_value second_value third_value fourth_value  \
0    bin_0   int64            0            0           0            0   
1    bin_1   int64            0            1           0            1   
2    bin_2   int64            0            0           0            0   
3    bin_3  object            T            T           F            F   
4    bin_4  object            Y            Y           Y            Y   
5    nom_0  object        Green        Green        Blue          Red   
6    nom_1  object     Triangle    Trapezoid   Trapezoid    Trapezoid   
7    nom_2  object        Snake      Hamster        Lion        Snake   
8    nom_3  object      Finland       Russia      Russia       Canada   
9    nom_4  object      Bassoon        Piano    Theremin         Oboe   
10   nom_5  object    50f116bcf    b3b4d25d0   3263bdce5    f12246592   
11   nom_6  object    3ac1b8814    fbcb50fc1   0922e3cb8    50d7ad46a   
12   nom_7  object    68f6ad3e9    3b6dd5612   a6a36f527    ec69236eb   
13   nom_8  object    c389000ab    4cd920251   de9c9f684    4ade6ab69   
14   nom_9  object    2f4cb3d51    f83c56c21   ae6800dd0    8270f0d71   
15   ord_0   int64            2            1           1            1   
16   ord_1  object  Grandmaster  Grandmaster      Expert  Grandmaster   
17   ord_2  object         Cold          Hot    Lava Hot  Boiling Hot   
18   ord_3  object            h            a           h            i   
19   ord_4  object            D            A           R            D   
20   ord_5  object           kr           bF          Jc           kW   
21     day   int64            2            7           7            2   
22   month   int64            2            8           2            1   
23  target   int64            0            0           0            1   

    fifth_value  eigenvalue_count  null_count  
0             0                 2           0  
1             0                 2           0  
2             0                 2           0  
3             F                 2           0  
4             N                 2           0  
5           Red                 3           0  
6     Trapezoid                 6           0  
7          Lion                 6           0  
8        Canada                 6           0  
9          Oboe                 4           0  
10    5b0f5acd5               222           0  
11    1fe17a1fd               522           0  
12    04ddac2be              1220           0  
13    cb43ab175              2215           0  
14    b164b72a7             11981           0  
15            1                 3           0  
16  Grandmaster                 5           0  
17     Freezing                 6           0  
18            a                15           0  
19            R                26           0  
20           qP               192           0  
21            7                 7           0  
22            8                12           0  
23            0                 2           0

테스트 세트에는 'target' 피처가 없습니다. 'target' 피처를 제거하겠습니다.

In [12]:
all_data = all_data.drop('target', axis=1)

## **2.2. bin_3, bin_4**

### **2.2.1. 매핑**

'bin_3', 'bin_4' 피처의 각 값을 숫자로 매핑하겠습니다.

In [13]:
all_data['bin_3'] = all_data['bin_3'].map({'T': 1, 'F': 0})
all_data['bin_4'] = all_data['bin_4'].map({'Y': 1, 'N': 0})

## **2.3. nom_num**

### **2.3.1. 원-핫 인코딩**

전체 'nom_num' 피처에 `OneHotEncoder` 클래스를 사용하여 원-핫 인코딩을 수행하겠습니다. 물론 (이 데이터에 한정하여) 판다스의 `get_dummies()` 메서드를 사용해도 무방하지만 `get_dummies()` 메서드는 결정적인 단점을 가집니다. 학습 세트의 피처를 학습하는 과정 없이 그대로 원-핫 인코딩을 수행하기 때문에 학습 세트에만 존재하고 테스트 세트에 없는 값은 원-핫 인코딩이 적용된 피처로 변환하지 않습니다. 게다가 `OneHotEncoder` 클래스보다 더 많은 메모리가 필요합니다. `OneHotEncoder` 클래스는 원-핫 인코딩을 수행하여 CSR 형식의 행렬을 반환하기 때문입니다(데이터프레임보다 적은 메모리 소모). 따라서 원-핫 인코딩을 적용하는 경우 습관적으로 `OneHotEncoder` 클래스를 사용하는 편이 좋습니다. [링크](https://psystat.tistory.com/136)에서 상세한 설명을 확인할 수 있습니다.

In [14]:
from sklearn.preprocessing import OneHotEncoder

# 전체 'nom_num' 피처를 담은 리스트를 생성합니다.
list_nom_num = ['nom_' + str(i) for i in range(10)]

one_hot_encoder = OneHotEncoder()
# 학습 및 테스트 세트를 하나로 합쳤으므로 `fit_transform()` 메서드를 사용합니다.
# 학습 및 테스트 세트로 나뉜 상태라면 테스트 세트에는 `transform()` 메서드만 사용합니다.
nom_num_one_hot_encoded = one_hot_encoder.fit_transform(all_data[list_nom_num])

In [15]:
nom_num_one_hot_encoded

<500000x16276 sparse matrix of type '<class 'numpy.float64'>'
	with 5000000 stored elements in Compressed Sparse Row format>

## **2.4. ord_num**

### **2.4.1. 매핑**

순서형 피처는 값 순서에 핵심 의미를 가집니다. 'ord_1', 'ord_2' 피처의 고윳값을 확인해 보겠습니다.

In [16]:
list_ord_1_2 = ['ord_1', 'ord_2']

for feature in list_ord_1_2:
    print(f'{feature}:\n {all_data[feature].unique()}\n')

ord_1:
 ['Grandmaster' 'Expert' 'Novice' 'Contributor' 'Master']

ord_2:
 ['Cold' 'Hot' 'Lava Hot' 'Boiling Hot' 'Freezing' 'Warm']



값 순서에 알맞도록 숫자로 매핑하겠습니다.

In [17]:
value_sequence_ord_1 = {'Novice':0, 'Contributor': 1, 'Expert': 2,
                  'Master': 3, 'Grandmaster': 4}
value_sequence_ord_2 = {'Freezing': 0, 'Cold': 1, 'Warm': 2, 'Hot': 3,
                  'Boiling Hot': 4, 'Lava Hot': 5}

all_data['ord_1'] = all_data['ord_1'].map(value_sequence_ord_1)
all_data['ord_2'] = all_data['ord_2'].map(value_sequence_ord_2)

'ord_3 ~ 5' 피처의 고윳값을 확인해 보겠습니다.

In [18]:
list_ord_3_4_5 = ['ord_3', 'ord_4', 'ord_5']

for feature in list_ord_3_4_5:
    print(f'{feature}:\n {all_data[feature].unique()}\n')

ord_3:
 ['h' 'a' 'i' 'j' 'g' 'e' 'd' 'b' 'k' 'f' 'l' 'n' 'o' 'c' 'm']

ord_4:
 ['D' 'A' 'R' 'E' 'P' 'K' 'V' 'Q' 'Z' 'L' 'F' 'T' 'U' 'S' 'Y' 'B' 'H' 'J'
 'N' 'G' 'W' 'I' 'O' 'C' 'X' 'M']

ord_5:
 ['kr' 'bF' 'Jc' 'kW' 'qP' 'PZ' 'wy' 'Ed' 'qo' 'CZ' 'qX' 'su' 'dP' 'aP'
 'MV' 'oC' 'RL' 'fh' 'gJ' 'Hj' 'TR' 'CL' 'Sc' 'eQ' 'kC' 'qK' 'dh' 'gM'
 'Jf' 'fO' 'Eg' 'KZ' 'Vx' 'Fo' 'sV' 'eb' 'YC' 'RG' 'Ye' 'qA' 'lL' 'Qh'
 'Bd' 'be' 'hT' 'lF' 'nX' 'kK' 'av' 'uS' 'Jt' 'PA' 'Er' 'Qb' 'od' 'ut'
 'Dx' 'Xi' 'on' 'Dc' 'sD' 'rZ' 'Uu' 'sn' 'yc' 'Gb' 'Kq' 'dQ' 'hp' 'kL'
 'je' 'CU' 'Fd' 'PQ' 'Bn' 'ex' 'hh' 'ac' 'rp' 'dE' 'oG' 'oK' 'cp' 'mm'
 'vK' 'ek' 'dO' 'XI' 'CM' 'Vf' 'aO' 'qv' 'jp' 'Zq' 'Qo' 'DN' 'TZ' 'ke'
 'cG' 'tP' 'ud' 'tv' 'aM' 'xy' 'lx' 'To' 'uy' 'ZS' 'vy' 'ZR' 'AP' 'GJ'
 'Wv' 'ri' 'qw' 'Xh' 'FI' 'nh' 'KR' 'dB' 'BE' 'Bb' 'mc' 'MC' 'tM' 'NV'
 'ih' 'IK' 'Ob' 'RP' 'dN' 'us' 'dZ' 'yN' 'Nf' 'QM' 'jV' 'sY' 'wu' 'SB'
 'UO' 'Mx' 'JX' 'Ry' 'Uk' 'uJ' 'LE' 'ps' 'kE' 'MO' 'kw' 'yY' 'zU' 'bJ'
 'Kf' 'ck' 'mb' 'Os' 'Ps

고윳값 개수가 많기 때문에 일일이 매핑하기는 번거롭습니다. `OrdinalEncoder` 클래스를 활용하여 값 순서에 맞게 숫자로 매핑하겠습니다.

In [19]:
from sklearn.preprocessing import OrdinalEncoder

ord_encoder = OrdinalEncoder()
all_data[list_ord_3_4_5] = ord_encoder.fit_transform(all_data[list_ord_3_4_5])

아래와 같이 값 순서가 `categories_` 속성에 저장되고, 각 값은 값 순서에 맞게 숫자로 매핑됩니다.

In [20]:
ord_encoder.categories_

[array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
        'n', 'o'], dtype=object),
 array(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
        'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'],
       dtype=object),
 array(['AP', 'Ai', 'Aj', 'BA', 'BE', 'Bb', 'Bd', 'Bn', 'CL', 'CM', 'CU',
        'CZ', 'Cl', 'DH', 'DN', 'Dc', 'Dx', 'Ed', 'Eg', 'Er', 'FI', 'Fd',
        'Fo', 'GD', 'GJ', 'Gb', 'Gx', 'Hj', 'IK', 'Id', 'JX', 'Jc', 'Jf',
        'Jt', 'KR', 'KZ', 'Kf', 'Kq', 'LE', 'MC', 'MO', 'MV', 'Mf', 'Ml',
        'Mx', 'NV', 'Nf', 'Nk', 'OR', 'Ob', 'Os', 'PA', 'PQ', 'PZ', 'Ps',
        'QM', 'Qb', 'Qh', 'Qo', 'RG', 'RL', 'RP', 'Rm', 'Ry', 'SB', 'Sc',
        'TR', 'TZ', 'To', 'UO', 'Uk', 'Uu', 'Vf', 'Vx', 'WE', 'Wc', 'Wv',
        'XI', 'Xh', 'Xi', 'YC', 'Yb', 'Ye', 'ZR', 'ZS', 'Zc', 'Zq', 'aF',
        'aM', 'aO', 'aP', 'ac', 'av', 'bF', 'bJ', 'be', 'cA', 'cG', 'cW',
        'ck', 'cp', 'dB', 'dE', 'dN', 'dO', 'dP', 'dQ', 'd

In [21]:
for feature in list_ord_3_4_5:
    print(f'{feature}:\n {all_data[feature].unique()}\n')

ord_3:
 [ 7.  0.  8.  9.  6.  4.  3.  1. 10.  5. 11. 13. 14.  2. 12.]

ord_4:
 [ 3.  0. 17.  4. 15. 10. 21. 16. 25. 11.  5. 19. 20. 18. 24.  1.  7.  9.
 13.  6. 22.  8. 14.  2. 23. 12.]

ord_5:
 [136.  93.  31. 134. 158.  53. 185.  17. 160.  11. 159. 170. 105.  90.
  41. 147.  60. 116. 117.  27.  66.   8.  65. 110. 129. 157. 108. 118.
  32. 115.  18.  35.  73.  22. 167. 111.  80.  59.  82. 155. 139.  57.
   6.  95. 120. 138. 145. 131.  92. 175.  33.  51.  19.  56. 151. 178.
  16.  79. 152.  15. 166. 163.  71. 169. 190.  25.  37. 106. 122. 132.
 127.  10.  21.  52.   7. 114. 121.  91. 165. 102. 148. 150. 100. 144.
 181. 113. 104.  77.   9.  72.  89. 161. 128.  86.  58.  14.  67. 135.
  97. 172. 176. 173.  88. 187. 141.  68. 180.  84. 183.  83.   0.  24.
  76. 164. 162.  78.  20. 146.  34. 101.   4.   5. 143.  39. 171.  45.
 124.  28.  49.  61. 103. 177. 107. 188.  46.  55. 126. 168. 184.  64.
  69.  44.  30.  63.  70. 174.  38. 154. 130.  40. 137. 189. 191.  94.
  36.  99. 142.  50.  54

### **2.4.2. min-max 정규화**

min-max 정규화를 수행하겠습니다.

In [22]:
from sklearn.preprocessing import MinMaxScaler

list_ord_num = ['ord_' + str(i) for i in range(6)]

min_max_scaler = MinMaxScaler()
all_data[list_ord_num] = min_max_scaler.fit_transform(all_data[list_ord_num])

In [23]:
for feature in list_ord_num:
    print(f'{feature}:\n {all_data[feature].unique()}\n')

ord_0:
 [0.5 0.  1. ]

ord_1:
 [1.   0.5  0.   0.25 0.75]

ord_2:
 [0.2 0.6 1.  0.8 0.  0.4]

ord_3:
 [0.5        0.         0.57142857 0.64285714 0.42857143 0.28571429
 0.21428571 0.07142857 0.71428571 0.35714286 0.78571429 0.92857143
 1.         0.14285714 0.85714286]

ord_4:
 [0.12 0.   0.68 0.16 0.6  0.4  0.84 0.64 1.   0.44 0.2  0.76 0.8  0.72
 0.96 0.04 0.28 0.36 0.52 0.24 0.88 0.32 0.56 0.08 0.92 0.48]

ord_5:
 [0.71204188 0.48691099 0.16230366 0.70157068 0.82722513 0.27748691
 0.96858639 0.08900524 0.83769634 0.05759162 0.83246073 0.89005236
 0.54973822 0.47120419 0.21465969 0.76963351 0.31413613 0.60732984
 0.61256545 0.14136126 0.34554974 0.04188482 0.34031414 0.57591623
 0.67539267 0.82198953 0.56544503 0.61780105 0.16753927 0.60209424
 0.09424084 0.18324607 0.38219895 0.11518325 0.87434555 0.58115183
 0.41884817 0.30890052 0.42931937 0.81151832 0.72774869 0.29842932
 0.03141361 0.4973822  0.62827225 0.72251309 0.7591623  0.68586387
 0.48167539 0.91623037 0.17277487 0.267015

## **2.5. day, month**

'day', 'month' 피처에는 원-핫 인코딩을 적용하겠습니다. 순서형 피처로 간주하여 min-max 정규화를 적용해 볼 수도 있습니다.

In [24]:
list_date = ['day', 'month']

date_one_hot_encoded = one_hot_encoder.fit_transform(all_data[list_date])

In [25]:
date_one_hot_encoded

<500000x19 sparse matrix of type '<class 'numpy.float64'>'
	with 1000000 stored elements in Compressed Sparse Row format>

## **2.6. 불필요한 피처 제거**

대부분의 피처 엔지니어링을 마쳤습니다. 원-핫 인코딩을 적용한 피처들은 CSR 형식으로 따로 저장된 상태이므로 기존 'nom_num', 'day', 'month' 피처는 불필요합니다. 삭제하겠습니다.

In [26]:
list_nom_num = ['nom_' + str(i) for i in range(10)]
list_date = ['day', 'month']

for features in [list_nom_num, list_date]:
    all_data = all_data.drop(features, axis=1)

## **2.7. 데이터 세트와 인코딩된 피처 간 결합**

데이터 세트와 인코딩된 피처를 결합하겠습니다. 기존 데이터를 CSR 형식으로 변환하여 인코딩된 CSR 행렬과 결합하고, 메모리를 절약하기 위해 CSR 형식을 그대로 유지시키겠습니다. 

In [27]:
from scipy import sparse

all_data_sparse = sparse.hstack([sparse.csr_matrix(all_data), # 기존 데이터를 CSR 형식으로 변환합니다.
                                 nom_num_one_hot_encoded,
                                 date_one_hot_encoded], format='csr')

In [28]:
all_data_sparse

<500000x16306 sparse matrix of type '<class 'numpy.float64'>'
	with 9163718 stored elements in Compressed Sparse Row format>

# **3. 첫 번째 모델링 및 하이퍼파라미터 튜닝**

하이퍼파라미터 최적화를 통해 성능을 개선해 보겠습니다.

## **3.1. 데이터 준비**

데이터를 준비하겠습니다.

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

train_num = len(train)
X_train_sparse = all_data_sparse[:train_num]
X_test_sparse = all_data_sparse[train_num:]
y_train_original = train['target']

X_train, X_valid, y_train, y_valid = train_test_split(
    X_train_sparse, y_train_original, test_size=0.15, stratify=y_train_original, random_state=2023)

## **3.2. XGBoost**

### **3.2.1. optuna**

`optuna` 프레임워크를 사용하겠습니다.

In [30]:
pip install optuna

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 365.3/365.3 KB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 210.6/210.6 KB 11.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.7/78.7 KB 8.0 MB/s eta 0:00:00


In [31]:
import optuna
from optuna.samplers import TPESampler

# 시도 과정을 출력하지 않는 코드입니다.
# optuna.logging.set_verbosity(optuna.logging.WARNING)

첫 번째 성능 개선 챕터와 달리 `'tree_method': 'gpu_hist'`로 지정하여 GPU와 협력하면서 하이퍼파라미터를 탐색하겠습니다. 아울러 학습 시간을 단축하기 위해 모델 성능 향상과 관련된 핵심적인 파라미터 몇 가지만 중점적으로 탐색해 보겠습니다. 탐색 횟수는 넉넉히 500회로 지정하겠습니다. 탐색 횟수가 높을수록 최적 하이퍼파라미터를 찾을 가능성이 높아지기 때문입니다. 캐글 노트북에서 별도로 진행하겠습니다.

In [41]:
# `optuna`를 수행할 기반(방침)이 되는 함수를 생성합니다.
def objective_XGB(trial, X_train, y_train, X_valid, y_valid):
    # 최적화할 하이퍼파라미터를 지정합니다.
    params = {
        'booster': 'gbtree',
        'objective': 'binary:logistic',
        'learning_rate': trial.suggest_float('learning_rate', 0.0001, 0.1),
        'n_estimators': trial.suggest_int('n_estimators', 100, 2000),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'scale_pos_weight': 2.269, # 타깃값 0의 개수를 타깃값 1의 개수로 나눈 값입니다.

        'n_jobs': -1,
        'random_state': 2023
    }

    # 모델 학습 후 roc auc까지 도출합니다.
    model = XGBClassifier(**params)
    model.fit(X_train, y_train, eval_set=[(X_valid, y_valid)],
              early_stopping_rounds=50, verbose=False)
    prediction = model.predict_proba(X_valid)[:, 1]
    roc_auc = roc_auc_score(y_valid, prediction)

    return roc_auc

In [ ]:
# 적용할 파라미터 목록입니다.
    # params = {
    #     'booster': trial.suggest_categorical('booster', ['gbtree', 'dart']),
    #     'objective': 'binary:logistic',
    #     'learning_rate': trial.suggest_float('learning_rate', 0.0001, 0.1),
    #     'n_estimators': trial.suggest_int('n_estimators', 100, 2000),
    #     'max_depth': trial.suggest_int('max_depth', 3, 10),
    #     'subsample': trial.suggest_float('subsample', 0.6, 1),
    #     'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1),
    #     'reg_alpha': trial.suggest_float('reg_alpha', 0, 5),
    #     'reg_lambda': trial.suggest_float('reg_lambda', 1.1, 1.5),
    #     'gamma': trial.suggest_float('gamma', 0, 8),
    #     'min_child_weight': trial.suggest_float('min_child_weight', 1, 2),
    #     'scale_pos_weight': 2.269, # 타깃값 0의 개수를 타깃값 1의 개수로 나눈 값입니다.
    #     'tree_method': 'gpu_hist', # 데이터 크기가 크므로 GPU를 사용합니다.
    #     'n_jobs': -1,
    #     'random_state': 2023
    # }

In [ ]:
# roc auc가 최대가 되는 값을 추구합니다.
study_xgb = optuna.create_study(direction='maximize', sampler=TPESampler())
# 최적 하이퍼파라미터를 500회만큼 탐색합니다.
study_xgb.optimize(lambda trial: objective_XGB(trial, X_train, y_train, X_valid, y_valid),
                   n_trials=500, show_progress_bar=True)

print(f'Best trial score: {study_xgb.best_trial.value}') # 가장 좋은 점수 및 시도 차례를 출력합니다.
print(f'Best params: {study_xgb.best_trial.params}') # 최적 하이퍼파라미터를 출력합니다.

[I 2023-01-22 20:43:42,609] A new study created in memory with name: no-name-162c3b15-3f8b-4282-a384-6bc70f8431b2
/usr/local/lib/python3.8/dist-packages/optuna/progress_bar.py:56: ExperimentalWarning: Progress bar is experimental (supported from v1.2.0). The interface can change in the future.
  self._init_valid()


  0%|          | 0/500 [00:00<?, ?it/s]

[I 2023-01-22 20:54:21,171] Trial 0 finished with value: 0.7781788682480579 and parameters: {'learning_rate': 0.09763321719869399, 'n_estimators': 1461, 'max_depth': 10}. Best is trial 0 with value: 0.7781788682480579.
[I 2023-01-22 21:00:30,668] Trial 1 finished with value: 0.7764150431808445 and parameters: {'learning_rate': 0.05415999027625729, 'n_estimators': 1290, 'max_depth': 9}. Best is trial 0 with value: 0.7781788682480579.
[I 2023-01-22 21:02:03,337] Trial 2 finished with value: 0.7658124937618086 and parameters: {'learning_rate': 0.04123974491690409, 'n_estimators': 344, 'max_depth': 3}. Best is trial 0 with value: 0.7781788682480579.
[I 2023-01-22 21:06:59,720] Trial 3 finished with value: 0.7757531668869432 and parameters: {'learning_rate': 0.08132461982720901, 'n_estimators': 1155, 'max_depth': 10}. Best is trial 0 with value: 0.7781788682480579.


소요시간 25시간 53분까지 탐색한 결과는 아래와 같습니다.

```
[I 2023-01-22 05:53:13,124] Trial 58 finished with value: 0.7883109672519307 and parameters: {'learning_rate': 0.09786587063265861, 'n_estimators': 9795, 'max_depth': 5}. Best is trial 57 with value: 0.7937624923748361.
```

총 58회만큼 탐색했으며 57회에 발견한 최적 하이퍼파라미터를 적용하여 훈련한 점수는 0.7938입니다. 베이스라인 모델링 챕터에서 LogisticRegression으로 하이퍼파라미터 튜닝 작업 없이 그대로 적용한 결과는 0.7996입니다. XGBoost가 약간 낮은 수준입니다. 연습 문제이므로 더 공들이지 않고 중단하겠습니다. 지나치게 오랜 시간이 소요됩니다.

탐색을 통해 찾은 하이퍼파라미터를 기반으로 전체 데이터를 사용하여 모델을 훈련하겠습니다. 물론 데이터 전체를 적용하여 탐색한 하이퍼파라미터가 아니기 때문에 전체 데이터에 알맞은 최적 하이퍼파라미터는 아닐 수 있습니다. 그렇다고 큰 차이까지 나진 않습니다.

In [31]:
model_xgb_optuna = XGBClassifier(
    max_depth=5,
    learning_rate=0.09786587063265861,
    n_estimators=9795,
    objective='binary:logistic',
    booster='gbtree',
    n_jobs=-1,
    scale_pos_weight=2.269,
    random_state=2023
    )
model_xgb_optuna.fit(X_train_sparse, y_train_original)

XGBClassifier(learning_rate=0.09786587063265861, max_depth=5, n_estimators=9795,
              n_jobs=-1, random_state=2023, scale_pos_weight=2.269)

In [32]:
submission['target'] = model_xgb_optuna.predict_proba(X_test_sparse)[:, 1]
# submission.to_csv('Categorical_Feature_Encoding_Challenge_submission8.csv')

프라이빗 스코어는 0.78196입니다. 기본 하이퍼파라미터를 사용한 모델보다는 개선됐지만 만족스럽지 않은 성능입니다.

## **3.3. LightGBM**

### **3.3.1. optuna**

LightGBM은 XGBoost보다는 수행 시간이 덜 들지만 역시 만만치 않습니다. 많은 하이퍼파라미터를 탐색하되 반복 횟수를 50회로 줄여서 탐색을 수행하겠습니다.

In [ ]:
def objective_LGBM(trial, X_train, y_train, X_valid, y_valid):
    params = {
        'boosting_type': trial.suggest_categorical('boosting_type', ['gbdt', 'dart']),
        'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.1),
        'n_estimators': trial.suggest_int('n_estimators', 100, 10000),
        'num_leaves': trial.suggest_int('num_leaves', 30, 40),
        'max_depth': -1,
        'subsample': trial.suggest_float('subsample', 0.6, 1),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1),
        'reg_alpha': trial.suggest_float('reg_alpha', 0.7, 0.9),
        'reg_lambda': trial.suggest_float('reg_lambda', 0.9, 1),
        'min_child_samples': trial.suggest_int('min_child_samples', 6, 10),
        'min_child_weight': trial.suggest_float('min_child_weight', 10, 40),
        'random_state': 2023
    }

    model = LGBMClassifier(**params)
    model.fit(X_train, y_train, eval_set=[(X_valid, y_valid)],
              early_stopping_rounds=50, verbose=False)
    prediction = model.predict_proba(X_valid)[:, 1]
    roc_auc = roc_auc_score(y_valid, prediction)

    return roc_auc

In [ ]:
study_lgbm = optuna.create_study(direction='maximize', sampler=TPESampler())
study_lgbm.optimize(lambda trial: objective_LGBM(trial, X_train, y_train, X_valid, y_valid),
                    n_trials=50, show_progress_bar=True)

print(f'Best trial score: {study_lgbm.best_trial.value}')
print(f'Best params: {study_lgbm.best_trial.params}')

[I 2023-01-20 09:32:14,593] A new study created in memory with name: no-name-a120530f-7164-4300-be24-8d36f2e06220
/usr/local/lib/python3.8/dist-packages/optuna/progress_bar.py:56: ExperimentalWarning: Progress bar is experimental (supported from v1.2.0). The interface can change in the future.
  self._init_valid()


  0%|          | 0/50 [00:00<?, ?it/s]

[I 2023-01-20 09:33:28,807] Trial 0 finished with value: 0.7898965594295287 and parameters: {'learning_rate': 0.07956369894405894, 'n_estimators': 2608, 'num_leaves': 32, 'subsample': 0.6265924109019706, 'colsample_bytree': 0.8031282581540565, 'reg_alpha': 0.7156150567409125, 'reg_lambda': 0.9339930839510621, 'min_child_samples': 8, 'min_child_weight': 38.71266335179969}. Best is trial 0 with value: 0.7898965594295287.
[I 2023-01-20 09:34:53,329] Trial 1 finished with value: 0.7904565457674677 and parameters: {'learning_rate': 0.0733661119447559, 'n_estimators': 7057, 'num_leaves': 35, 'subsample': 0.7348837513430502, 'colsample_bytree': 0.8680469013982914, 'reg_alpha': 0.8702023037296185, 'reg_lambda': 0.9809488915749904, 'min_child_samples': 10, 'min_child_weight': 35.90811953176664}. Best is trial 1 with value: 0.7904565457674677.
[I 2023-01-20 09:36:03,846] Trial 2 finished with value: 0.7910656936615447 and parameters: {'learning_rate': 0.08209502738335393, 'n_estimators': 9487, '

탐색 34회에 최적 하이퍼파라미터를 찾았습니다.

전체 데이터를 훈련에 사용하겠습니다.

In [ ]:
params = study_lgbm.best_trial.params
model_lgbm_optuna = LGBMClassifier(**params)
model_lgbm_optuna.fit(X_train_sparse, y_train_original)

LGBMClassifier(colsample_bytree=0.8747692098547708,
               learning_rate=0.07401073362642446, min_child_samples=8,
               min_child_weight=13.444943166244059, n_estimators=6997,
               reg_alpha=0.8396867483808421, reg_lambda=0.9775508631575757,
               subsample=0.8016577934815098)

파일을 제출하겠습니다.

In [ ]:
submission['target'] = model_lgbm_optuna.predict_proba(X_test_sparse)[:, 1]
# submission.to_csv('Categorical_Feature_Encoding_Challenge_submission9.csv')

프라이빗 스코어는 0.78439입니다. 기본 하이퍼파라미터를 사용한 모델보다는 성능이 개선되었고 XGBoost 모델보다 높은 성능이지만, 도리어 LogisticRegression 모델보다 성능이 떨어집니다.

## **3.4. LogisticRegression**

### **3.4.1. optuna**

하이퍼파라미터 탐색 범위를 늘리고 반복 횟수도 500회로 늘려 보겠습니다.

In [32]:
def objective_LogisticRegression(trial, X_train, y_train, X_valid, y_valid):
    params = {
        'C': trial.suggest_float('C', 0.001, 1000),
        'max_iter': trial.suggest_int('max_iter', 100, 10000),
        'solver': 'liblinear',
        'random_state': 2023
    }

    model = LogisticRegression(**params)
    model.fit(X_train, y_train)
    prediction = model.predict_proba(X_valid)[:, 1]
    roc_auc = roc_auc_score(y_valid, prediction)

    return roc_auc

In [33]:
study_logistic = optuna.create_study(direction='maximize', sampler=TPESampler())
study_logistic.optimize(lambda trial: objective_LogisticRegression(trial, X_train, y_train, X_valid, y_valid),
                        n_trials=500, show_progress_bar=True)

print(f'Best trial score: {study_logistic.best_trial.value}')
print(f'Best params: {study_logistic.best_trial.params}')

[I 2023-01-22 09:28:33,531] A new study created in memory with name: no-name-1ba5420e-9b22-42e4-a00a-fe6f4dec5135
/usr/local/lib/python3.8/dist-packages/optuna/progress_bar.py:56: ExperimentalWarning: Progress bar is experimental (supported from v1.2.0). The interface can change in the future.
  self._init_valid()


  0%|          | 0/500 [00:00<?, ?it/s]

[I 2023-01-22 09:30:06,940] Trial 0 finished with value: 0.7846213153617769 and parameters: {'C': 908.5168091037731, 'max_iter': 929}. Best is trial 0 with value: 0.7846213153617769.
[I 2023-01-22 09:30:57,466] Trial 1 finished with value: 0.787009668745611 and parameters: {'C': 63.066799853180896, 'max_iter': 7130}. Best is trial 1 with value: 0.787009668745611.
[I 2023-01-22 09:32:33,429] Trial 2 finished with value: 0.7845043412338104 and parameters: {'C': 825.7711839267548, 'max_iter': 429}. Best is trial 1 with value: 0.787009668745611.
[I 2023-01-22 09:33:25,507] Trial 3 finished with value: 0.7862322221240842 and parameters: {'C': 222.2592241193612, 'max_iter': 8513}. Best is trial 1 with value: 0.787009668745611.
[I 2023-01-22 09:34:41,041] Trial 4 finished with value: 0.7856695866395093 and parameters: {'C': 195.99274614734236, 'max_iter': 7970}. Best is trial 1 with value: 0.787009668745611.
[I 2023-01-22 09:35:35,539] Trial 5 finished with value: 0.7882955026549322 and param

탐색 231회에 최적 하이퍼파라미터를 찾았습니다. 최고 점수인 0.8053205237751317보다 0.00002만큼 낮은 수준입니다. 성능이 매우 미세하게 떨어질 것으로 예상합니다. 전체 데이터를 사용하여 모델을 훈련해 보겠습니다.

In [34]:
params = study_logistic.best_trial.params
model_logistic = LogisticRegression(**params)
model_logistic.fit(X_train_sparse, y_train_original)

LogisticRegression(C=0.09846929569930918, max_iter=1898)

파일을 제출해 보겠습니다.

In [35]:
submission['target'] = model_logistic.predict_proba(X_test_sparse)[:, 1]
submission.to_csv('Categorical_Feature_Encoding_Challenge_submission9.csv')

프라이빗 스코어는 0.80268입니다. 추측대로 이전 스코어인 0.80275보다 미세하게 떨어졌습니다. 순위 또한 15위에서 19위로 떨어졌습니다. 기실 실무에서 이 정도 성능 저하는 큰 문제가 되지 않습니다. `optuna` 패키지가 최적 파라미터를 잘 탐색했다고 평가할 만한 수준입니다.

# **4. 두 번째 모델링 및 하이퍼파라미터 튜닝**

## **4.1. 데이터 준비**

학습 및 검증 세트 비율을 0.15에서 0.1로 축소하여 하이퍼파라미터 최적화를 다시 한번 수행해 보겠습니다. 현재까지 가장 좋은 성능을 보였던 로지스틱 회귀 모델의 최적 하이퍼파라미터만 탐색하겠습니다.

In [32]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X_train_sparse, y_train_original, test_size=0.1, stratify=y_train_original, random_state=2023)

## **4.2. LogisticRegression**

### **4.2.1. optuna**

In [37]:
def objective_LogisticRegression(trial, X_train, y_train, X_valid, y_valid):
    params = {
        'C': trial.suggest_float('C', 0.1, 10),
        'max_iter': trial.suggest_int('max_iter', 100, 10000),
        'solver': 'liblinear',
        'random_state': 2023
    }

    model = LogisticRegression(**params)
    model.fit(X_train, y_train)
    prediction = model.predict_proba(X_valid)[:, 1]
    roc_auc = roc_auc_score(y_valid, prediction)

    return roc_auc

In [38]:
study_logistic = optuna.create_study(direction='maximize', sampler=TPESampler())
study_logistic.optimize(lambda trial: objective_LogisticRegression(trial, X_train, y_train, X_valid, y_valid),
                        n_trials=500, show_progress_bar=True)

print(f'Best trial score: {study_logistic.best_trial.value}')
print(f'Best params: {study_logistic.best_trial.params}')

[I 2023-01-22 18:36:21,923] A new study created in memory with name: no-name-fbb0da3e-54d0-4deb-9202-de293c1ffd40
/usr/local/lib/python3.8/dist-packages/optuna/progress_bar.py:56: ExperimentalWarning: Progress bar is experimental (supported from v1.2.0). The interface can change in the future.
  self._init_valid()


  0%|          | 0/500 [00:00<?, ?it/s]

[I 2023-01-22 18:36:56,247] Trial 0 finished with value: 0.7891953153862101 and parameters: {'C': 7.578692506041821, 'max_iter': 9286}. Best is trial 0 with value: 0.7891953153862101.
[I 2023-01-22 18:37:19,258] Trial 1 finished with value: 0.7942987002204887 and parameters: {'C': 1.662506209505343, 'max_iter': 1276}. Best is trial 1 with value: 0.7942987002204887.
[I 2023-01-22 18:37:50,067] Trial 2 finished with value: 0.7885333867584884 and parameters: {'C': 9.84151812533324, 'max_iter': 2442}. Best is trial 1 with value: 0.7942987002204887.
[I 2023-01-22 18:38:19,891] Trial 3 finished with value: 0.7892453726854636 and parameters: {'C': 7.410787242838828, 'max_iter': 7746}. Best is trial 1 with value: 0.7942987002204887.
[I 2023-01-22 18:38:45,182] Trial 4 finished with value: 0.78979574654153 and parameters: {'C': 6.009764251362852, 'max_iter': 1171}. Best is trial 1 with value: 0.7942987002204887.
[I 2023-01-22 18:39:11,241] Trial 5 finished with value: 0.7902620670485836 and par


탐색 211회에 최적 하이퍼파라미터를 찾았습니다. 최고 점수인 0.8053205237751317보다 0.002만큼 낮은 수준입니다. 성능이 매우 미세하게 떨어질 것으로 예상합니다. 전체 데이터를 사용하여 모델을 훈련해 보겠습니다.

In [39]:
params = study_logistic.best_trial.params
model_logistic = LogisticRegression(**params)
model_logistic.fit(X_train_sparse, y_train_original)

LogisticRegression(C=0.10834540601334458, max_iter=5585)

파일을 제출해 보겠습니다.

In [40]:
submission['target'] = model_logistic.predict_proba(X_test_sparse)[:, 1]
submission.to_csv('Categorical_Feature_Encoding_Challenge_submission11.csv')

예상과 달리 최고 점수인 0.80275보다 0.00001 높은 0.80276입니다. 15위에서 한 계단 오른 14위입니다.

# **5. 결론**

최종 모델링을 수행하는 다음 챕터에서 가장 좋은 성능을 내는 로지스틱 회귀 모델을 최종 모델로 선정하겠습니다. 일반적으로 앙상블 기반 결정트리 알고리즘은 로지스틱 회귀 알고리즘보다 좋은 성능을 냅니다. 결정적인 단점이라면 데이터 수가 많을 때 학습 시간이 매우 오래 걸립니다.

다시 강조하지만 실무에서는 모델 성능보다 우선해야 할 중요한 게 있습니다. 약간의 성능 저하를 감안하더라도 과(소)적합을 방지하는 조치가 훨씬 더 유의미하고 필수적입니다.

물론 대회에서는 매우 작은 점수 차로 순위가 뒤바뀝니다. 연습 대회가 아닌 실제 대회였다면 다른 알고리즘을 토대로 더 높은 점수를 획득하는 데 전념했을 것입니다.